### PR TITLE
Certificate Renewal in Connectivity Certs Controller

### PR DESCRIPTION
**Pull request by Szymongib**
**Description**

Connectivity Certs Controller can renew client certificates.

Changes proposed in this pull request:

- Added Central Connection Custom Resource
- Created a new Controller for Central Connections
- Certificates renewal implemented as a part of the new controller

**Related issue(s)**
#2931

### CODE REVIEW

File: components/connectivity-certs-controller/cmd/connectivitycertscontroller/options.go [hunk 1]
```diff
@@ -10,13 +10,17 @@ type options struct {
...
+	minimalSyncTime           int
 }
 
 func parseArgs() *options {
 	appName := flag.String("appName", "connectivity-certs-controller", "Name used in controller registration")
 	namespace := flag.String("namespace", "kyma-integration", "Namespace in which secrets are created")
 	clusterCertificatesSecret := flag.String("clusterCertificatesSecret", "cluster-client-certificates", "Secret name where cluster client certificate and key are kept")
 	caCertificatesSecret := flag.String("caCertificatesSecret", "ca-certificates", "Secret name where CA certificate is kept")
+	syncPeriod := flag.Int("syncPeriod", 300, "Time period between resyncing existing resources")
+	minimalSyncTime := flag.Int("minimalSyncTime", 300, "Minimal time between trying to synchronize with Central Connector Service")
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/connectivity-certs-controller/cmd/connectivitycertscontroller/options.go]: `syncPeriod` is corresponding to the mechanism of Kubernetes controller, which determines how often unchanged resources will be processed by the controller. `minimalSyncTime` on the other hand determines if the Connectivity Certs Controller should resynchronize connection with the central Connector Service. In my opinion, these parameters should be separated although to make it more clear we may consider renaming `minimalSyncTime` to something like `minimalConnectorSyncTime`.

File: components/connectivity-certs-controller/README.md [hunk 2]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
 EOF
 ```
 
 After a successful exchange of certificates, the controller creates new Secrets or modifies the existing ones that correspond to the updated certificates. The CertificateRequest CR is deleted.
 
+
+## Connection status
+
+When certificates are fetched successfully the `CentralConnection` CR is created.
+It contains the `ManagementInfoURL`, status of a connection with central the Connector Service and the certificate validity period.
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/connectivity-certs-controller/README.md]: Done

File: components/connectivity-certs-controller/internal/centralconnection/controller.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+type Controller struct {
+	masterConnectionClient  ResourceClient
+	certificatePreserver    certificates.Preserver
+	certificateProvider     certificates.Provider
+	mutualTLSClientProvider connectorservice.MutualTLSClientProvider
+	minimalSyncTime         time.Duration
+	logger                  *log.Entry
+}
+
+func InitCentralConnectionsController(
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/connectivity-certs-controller/internal/centralconnection/controller.go]: Done

File: components/connectivity-certs-controller/internal/centralconnection/controller.go [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+	certValidityRenewalThreshold = 0.3
+)
+
+type ResourceClient interface {
+	Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error
+	Update(ctx context.Context, obj runtime.Object) error
+}
+
+type Controller struct {
+	masterConnectionClient  ResourceClient
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/connectivity-certs-controller/internal/centralconnection/controller.go]: Done, it was left over after renaming

File: components/connectivity-certs-controller/internal/centralconnection/controller.go [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+func newCentralConnectionController(
+	client ResourceClient,
+	minimalSyncTime time.Duration,
+	certPreserver certificates.Preserver,
+	certProvider certificates.Provider,
+	mTLSClientProvider connectorservice.MutualTLSClientProvider) *Controller {
+
+	return &Controller{
+		masterConnectionClient:  client,
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/connectivity-certs-controller/internal/centralconnection/controller.go]: Done

File: components/connectivity-certs-controller/README.md [hunk 6]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
+```
+
+The Connectivity Certs Controller checks a status of the connection by calling `ManagementInfoURL` stored in `CentralConnection` CR.
+
+
+## Certificate renewal
+
+Connectivity Certs Controller renews client certificate when it is getting close to its expiration time. Renewed certificate will replace the previous one in the Secret together with new private key.
+
+To renew the certificate immediately `spec.RenewNow` field on `CentralConnection` should be set to `true`.
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/connectivity-certs-controller/README.md]: The behavior is described correctly as after updating `CentralConnection`, it is rescheduled by the controller immediately and therefore synchronized.

File: components/connectivity-certs-controller/README.md [hunk 7]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
 EOF
 ```
 
 After a successful exchange of certificates, the controller creates new Secrets or modifies the existing ones that correspond to the updated certificates. The CertificateRequest CR is deleted.
 
+
+## Connection status
+
+When certificates are fetched successfully the `CentralConnection` CR is created.
+It contains the `ManagementInfoURL`, status of a connection with the central Connector Service and the certificate validity period.
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
It contains **ManagementInfoURL**, the status of a connection with the central Connector Service, and the certificate validity period.
```

File: components/connectivity-certs-controller/README.md [hunk 8]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
+
+
+The `CentralConnection` CR is created with the same name as a name of `CertificateRequest` for which connection was established.
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
The CentralConnection CR is created with the same name as the name of `CertificateRequest` for which the connection was established.
```

File: components/connectivity-certs-controller/README.md [hunk 9]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
+To get `CentralConnection` run:
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
To get the CentralConnection CR, run:
```

File: components/connectivity-certs-controller/README.md [hunk 10]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
+```
+kubectl get centralconnections.applicationconnector.kyma-project.io {CENTRAL_CONNECTION_NAME} -oyaml
+```
+
+The example resource looks like that:
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
The example resource looks as follows:
```

File: components/connectivity-certs-controller/README.md [hunk 11]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
+status:
+  certificateStatus:
+    notAfter: "2019-07-11T08:47:38Z"
+    notBefore: "2019-04-10T08:47:38Z"
+  synchronizationStatus:
+    lastSuccess: "2019-04-10T09:01:06Z"
+    lastSync: "2019-04-10T09:01:06Z"
+```
+
+The Connectivity Certs Controller checks a status of the connection by calling `ManagementInfoURL` stored in `CentralConnection` CR.
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
The Connectivity Certs Controller checks the connection status by calling **ManagementInfoURL** stored in the CentralConnection CR.
```

File: components/connectivity-certs-controller/README.md [hunk 12]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
+
+
+## Certificate renewal
+
+Connectivity Certs Controller renews client certificate when it is getting close to its expiration time. Renewed certificate will replace the previous one in the Secret together with new private key.
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
The Connectivity Certs Controller renews a client certificate when it gets close to the expiration time. A renewed certificate will replace the previous one in the Secret together with a new private key.
```

File: components/connectivity-certs-controller/README.md [hunk 13]
```diff
@@ -14,14 +14,55 @@ cat <<EOF | kubectl apply -f -
...
+
+To renew the certificate immediately `spec.renewNow` field on `CentralConnection` should be set to `true`.
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
To renew the certificate, set the **spec.renewNow** field in the CentralConnection CR to `true`. 
```

File: components/connectivity-certs-controller/README.md [hunk 14]
```diff
@@ -30,3 +71,10 @@ To get the error message, run:
 ```
 kubectl get certificaterequests.applicationconnector.kyma-project.io {CERT_REQUEST_NAME} -o jsonpath={.status.error}
 ```
+
+Similarly if the synchronization with central Connector Service or certificate renewal fails, the error status will be set on `CentralConnection` CR.
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
Similarly, if the synchronization with the central Connector Service or the certificate renewal fails, the error status will be displayed in the CentralConnection CR.
```

File: components/connectivity-certs-controller/README.md [hunk 15]
```diff
@@ -30,3 +71,10 @@ To get the error message, run:
...
+
+To check it, run:
```
Comment (klaudiagrz) [inline components/connectivity-certs-controller/README.md]:
```suggestion
To check the error status, run:
```

**Pull request merged**